llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp

```diff
@@ -6855,6 +6855,47 @@ Instruction *InstCombinerImpl::visitICmpInst(ICmpInst &I) {
   if (Value *V = simplifyICmpInst(I.getPredicate(), Op0, Op1, Q))
     return replaceInstUsesWith(I, V);
 
+  {
+    Value *X;
+    const APInt *C1, *C2;
+    ICmpInst::Predicate Pred;
+    if ((match(&I, m_ICmp(Pred, m_SRem(m_Value(X), m_NonNegative(C1)),
+                          m_APInt(C2))) &&
+         *C2 == *C1 - 1 &&
+         (Pred == ICmpInst::ICMP_SLT || Pred == ICmpInst::ICMP_SGE)) ||
+        (match(&I,
+               m_ICmp(Pred, m_SRem(m_Value(X), m_Negative(C1)), m_APInt(C2))) &&
+         *C2 == *C1 + 1 &&
+         (Pred == ICmpInst::ICMP_SGT || Pred == ICmpInst::ICMP_SLE))) {
+      // icmp slt (X s% C), (C - 1) --> icmp ne (X s% C), (C - 1), if C >= 0
+      // icmp sge (X s% C), (C - 1) --> icmp eq (X s% C), (C - 1), if C >= 0
+      // icmp sgt (X s% C), (C + 1) --> icmp ne (X s% C), (C + 1), if C < 0
+      // icmp sle (X s% C), (C + 1) --> icmp eq (X s% C), (C + 1), if C < 0
+      auto *NewCmp = Builder.CreateICmp(
+          (Pred == ICmpInst::ICMP_SLT || Pred == ICmpInst::ICMP_SGT)
+              ? ICmpInst::ICMP_NE
+              : ICmpInst::ICMP_EQ,
+          Op0, Op1);
+      return replaceInstUsesWith(I, NewCmp);
+    }
+
+    if (match(&I, m_ICmp(Pred, m_URem(m_Value(X), m_NonNegative(C1)),
+                         m_APInt(C2))) &&
+        (Pred == ICmpInst::ICMP_SLT || Pred == ICmpInst::ICMP_SGE ||
+         Pred == ICmpInst::ICMP_UGE || Pred == ICmpInst::ICMP_ULT)) {
+      // icmp slt (X u% C), (C - 1) --> icmp ne (X u% C), (C - 1), if C >= 0
+      // icmp ult (X u% C), (C - 1) --> icmp eq (X u% C), (C - 1), if C >= 0
+      // icmp sge (X u% C), (C - 1) --> icmp eq (X u% C), (C - 1), if C >= 0
+      // icmp uge (X u% C), (C - 1) --> icmp ne (X u% C), (C - 1), if C >= 0
+      auto *NewCmp = Builder.CreateICmp(
+          (Pred == ICmpInst::ICMP_SLT || Pred == ICmpInst::ICMP_ULT)
+              ? ICmpInst::ICMP_NE
+              : ICmpInst::ICMP_EQ,
+          Op0, Op1);
+      return replaceInstUsesWith(I, NewCmp);
```

Comment:
```suggestion
      return CmpInst::Create(Instruction::ICmp, 
          (Pred == ICmpInst::ICMP_SLT || Pred == ICmpInst::ICMP_ULT)
              ? ICmpInst::ICMP_NE
              : ICmpInst::ICMP_EQ,
          Op0, Op1);
```
Use `XXXInst::Create` for the outermost inst.

Reply:
what is the difference between `replaceInstUsesWith` and `XXXInst::Create`?

Reply:
@dtcxzyw

Reply:
> what is the difference between `replaceInstUsesWith` and `XXXInst::Create`?

If you always create a new instructions rather than fold the pattern into a constant/an existing value, please use `XXXInst::Create` for the outermost inst.

`InstCombiner::replaceInstUsesWith` does more things to maintain the worklist. So using `XXXInst::Create` may improve the performance and reduce the binary size.